src/tree.rs

```diff
@@ -101,3 +101,138 @@ fn hash_layer(l_star: usize, pair: Pair) -> pallas::Base {
+    use std::convert::TryInto;
+
+    use crate::{
+        constants::{MERKLE_CRH_PERSONALIZATION, MERKLE_DEPTH_ORCHARD},
+        note::{commitment::ExtractedNoteCommitment, testing::arb_note, Note},
+        primitives::sinsemilla::{i2lebsp_k, HashDomain, K},
+        value::{testing::arb_positive_note_value, MAX_NOTE_VALUE},
+    };
+    use ff::{PrimeField, PrimeFieldBits};
+    use pasta_curves::pallas;
+
+    use proptest::collection::vec;
+    use proptest::prelude::*;
+
+    use super::{hash_layer, Anchor, MerklePath, Pair};
+
+    lazy_static! {
+        static ref EMPTY_ROOTS: Vec<pallas::Base> = {
+            // MerkleCRH Sinsemilla hash domain.
+            let domain = HashDomain::new(MERKLE_CRH_PERSONALIZATION);
+
+            let mut v = vec![pallas::Base::zero()];
+            for l_star in 0..MERKLE_DEPTH_ORCHARD {
+                let next = domain
+                .hash(
+                    std::iter::empty()
+                        .chain(i2lebsp_k(l_star).iter().copied().take(K))
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        )
+                        .chain(
+                            v[l_star]
+                            .to_le_bits()
+                            .iter()
+                            .by_val()
+                            .take(pallas::Base::NUM_BITS as usize)
+                        ),
+                )
+                .unwrap();
+                v.push(next);
+            }
+            v
+        };
+    }
+
+    prop_compose! {
+        /// Generates an arbitrary Merkle tree of with `n_notes` nonempty leaves.
+        pub fn arb_tree(n_notes: i32)
+        (
+            // generate note values that we're certain won't exceed MAX_NOTE_VALUE in total
+            notes in vec(
+                arb_positive_note_value(MAX_NOTE_VALUE / n_notes as u64).prop_flat_map(arb_note),
+                n_notes as usize
+            ),
+        )
+        -> (Vec<(Note, MerklePath)>, Anchor) {
+            // Inefficient algorithm to build a perfect subtree containing all notes.
+            let perfect_subtree_depth = (n_notes as f64).log2().ceil() as usize;
+            let commitments: Vec<ExtractedNoteCommitment> = notes.iter().map(|note| {
+                let cmx: ExtractedNoteCommitment = note.commitment().into();
+                cmx
+            }).collect();
+
+            let padded_leaves = {
+                let mut padded_leaves = commitments.clone();
+
+                let pad = (0..((1 << perfect_subtree_depth) - n_notes as usize)).map(
+                    |_| ExtractedNoteCommitment::uncommitted()
+                ).collect::<Vec<_>>();
+
+                padded_leaves.extend_from_slice(&pad);
+                padded_leaves
+            };
```

Comment:
Instead of representing the perfect tree with uncommitted leaves (since that API will be removed), use `Option<pallas::Base>`.